File: Chapter12/VB/HttpService/Teams/Connection/Receiver.vb
Imports System.Net.Sockets

Public Class Receiver
  Private _stopRequested As Boolean
  Private _socket As Socket

  Public Sub New(ByVal theSocket As Socket)
    _socket = theSocket
  End Sub

  Public Sub [Stop]()
    _stopRequested = True
  End Sub

  Public Sub Run()
    _stopRequested = False
    While Not _stopRequested
      CheckForIncomingTraffic()
    End While
  End Sub

  Sub CheckForIncomingTraffic()
    If _socket.Available = 0 Then
      FireIdle()
    Else
      FireResponse(_socket)
    End If
  End Sub

  Public Event OnResponse(ByVal theSocket As Socket)
  Sub FireResponse(ByVal theSocket As Socket)
    RaiseEvent OnResponse(theSocket)
  End Sub

  Public Event OnIdle()
  Sub FireIdle()
    RaiseEvent OnIdle()
  End Sub
End Class

File: Chapter12/C#/HttpService/Teams/Connection/Receiver.cs
using System;
using System.Net.Sockets;

namespace HttpService
{
	/// <summary>
	/// Receiver handles all the incoming traffic
	/// for a given socket
	/// </summary>
  public class Receiver
  {
    bool stopRequested;
    Socket socket;

	  public Receiver(Socket theSocket)
    {
      socket = theSocket;
    }

    public void Stop()
    {
      stopRequested = true;
    }

    public void Run()
    {
      stopRequested = false;
      while (!stopRequested)
        CheckForIncomingTraffic();
    }

    void CheckForIncomingTraffic()
    {
      if (socket.Available == 0)
        FireIdle();
      else
        FireResponse(socket);
    }

    public delegate void ResponseHandler(Socket theSocket);
    public event ResponseHandler OnResponse;
    void FireResponse(Socket theSocket)
    {
      if (OnResponse != null)
        OnResponse(theSocket);
    }

    public delegate void IdleHandler();
    public event IdleHandler OnIdle;
    void FireIdle()
    {
      if (OnIdle != null)
        OnIdle();
    }
  }
}
